### PR TITLE
ENH: speed up einsum with optimize using batched matmul

### PR DESCRIPTION
**Pull request by jcmgray**
This PR enables dispatching *all* pairwise calls in `einsum` when `optimize=` is turned on to batched matrix multiply (or non-batched `matmul` and `multiply` as special cases). For a short explanation of the implementation see here - https://github.com/jcmgray/einsum_bmm.

Currently falling back to `c_einsum` for these, when there exist batch indices, induces potentially many orders of magnitude slow downs (#22604, e.g. 7000x), particularly in the case of large or high dimensional tensors.

After fairly extensive benchmarking (below), I *believe* that this should be a uniform improvement across essentially all the cases when one would use `optimize=True`. This is partly enabled by the modern performance of matmul which in the past was not always faster than `einsum`. However, I know `einsum` has an enormous number of uses and I'd gladly accept more suggestions for benchmarks, and would understand people's hesitation! 

(@dgasmith, @seberg)

## Other notes:
- This replaces all the `can_blas` logic and `tensordot` calls, and simplifies everything into the `bmm_einsum`, which encapsulates how `tensordot` is implemented
- When there are no contracted indices, `numpy.multiply` is used for slightly better performance
- The current implementation can simply pass `out` and other kwargs on to `mutiply` and `matmul`, unlike tensordot, meaning `einsum('ij,jk->ik', x, y, optimize=True, out=z)` should write directly to `z`
- The fusing allows `einsum(eq, ..., optimize=True)` to be used when there are more than 32 indices involved (#5744).
- The low overhead of the actuall `bmm_einsum` implementation comes from caching what operations to use based on `(eq, x.shape, y.shape)`, using `@functools.lru_cache(2**12)`.
- the performance broadly speaking is brought in line with `torch.einsum`

## TODO: 
- [ ] update docs?
- [ ] decide cache size?
- [x] should I lint the existing `einsumfunc.py` code?

## Benchmarks

In the following the kernels are the following:

- `no optimize`: the base `np.einsum(eq, x, y)` i.e. `c_einsum`
- `optimize + dot`: the current `np.einsum(eq, x, y, optimize=True)` which calls `tensordot` where possible, and also induces some overhead from the potential path optimization (even though there are only two terms in the following, the equation must be checked and parsed etc.)
- `optimize + bmm`: the proposed `np.einsum(eq, x, y, optimize=True)` which calls batched matmul, and still induces some overhead from the potential path optimization
- `no optimize + bmm`: this calls the pairwise bmm einsum directly, and is provided simply to show what overhead comes from the bmm impl, and what from the potential path optimization 

The main two to compare are `optimize + dot` to `optimize + bmm`. Overall I'd summarize the results as the following:

1. when a contraction is memory bound, all methods are pretty much equivalent
2. when a contraction involves no batch dimensions, performance of `dot` is retained
3. when a contraction involves batch dimensions, performance can be orders of magnitude faster

All dimensions are size `n` unless otherwise noted.

vec_inner

vec_outer

square_matmul

square_matvec

square_vecmat

For `batch_matmul_small` the sizes are `(n, 2, 2), (n, 2, 2)`.

batch_matmul_small

For `batch_matmul_equal` the sizes are `(n, n, n), (n, n, n)`.

batch_matmul_equal

For `batch_matmul_large` the sizes are `(10, n, n), (10, n, n)`.

batch_matmul_large

hadamard

hadamard_unalinged

CCSDT_1

interleaved_dot

interleaved_batched_dot

In `random_extreme` the shapes are taken as the following:
```python
[
    (n, 4, 3, 4, 3, 4, 2, 4, 2, n, 2, n, 3, 2, 4),
    (2, 4, n, n, 4, 4, 4, 3, 4, 4, 4, 3, 4)
]
```

random_extreme

In `many_small_dims_dot` the shapes are `([2] * n, [2] * n)` and half in an interleaved pattern are contracted.

many_small_dims_dot

In `many_small_dims_batched_dot` the shapes are `([2] * n, [2] * n)` and a third in an interleaved pattern are contracted and a third in an interleaved pattern are batch indices.

many_small_dims_batched_dot

### CODE REVIEW

**Comment by jcmgray** [comment]
No rush on my part, but wondering if there are any thoughts about this change or things I can do for this PR?

**Comment by jcmgray** [comment]
Hi @seberg, ah yes I just checked and currently that matmul behaviour would result in a performance regression for the specific case that arrays reach the matmul call (after preparatory reshape/transpose etc) with whichever strides trigger the non-blas, no-copy call (#23588).

```python
import numpy as np
d = 1000
x = np.random.randn(d, d)
y = np.random.randn(d, d) + 1j * np.random.randn(d, d)

%%timeit
z = np.einsum('ij,jk->ik', x, y.real, optimize=True)
# -->
# 11.5 ms ± 473 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)  (current dot backend)
# -->
# 897 ms ± 23.7 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)  (new matmul backend)
```

If the solution in #23752 is to do the copy when needed then that is very much aligned with the logic of `optimize=True`, i.e. trade memory and copying for being able to BLAS-it. Else one could check in the einsum_bmm call but preferably it would be deferred to `matmul` I suppose.

**Comment by dgasmith** [comment]
> TBH, I had hoped @dgasmith can give a recommendation. The timings do look like this should work pretty generically, although its easy to miss cases, I guess.

It is hard for me to say, we benchmarked fairly throughly with `optimize=True`; however, the deployed result had a large number of edge cases that we missed. That being said, localizing the code to `optimize=True` is a good move to let everyone switch on/off performance.

**Comment by jcmgray** [comment]
Rather than wait on progress in https://github.com/numpy/numpy/issues/23588 would it be reasonable to simply require contiguity before the matmul in the `bmm_einsum` function here? 

I.e. do we expect there to be cases where the arrays are not contiguous, and we expect it to be faster to leave them so?

My guess is that this is only the case for certain small contractions that are not the focus of when people use `optimize=True`. 

Of course having some rough heuristic in `matmul` itself would be cleaner.

**Comment by ACE07-Sev** [comment]
@jcmgray Greetings dear Dr. Gray!

Pleasantly surprised to see you here. I was wondering, did this PR ever get merged?

**Comment by mattip** [comment]
It was not merged. `einsum` is quite complex, and we were waiting for anther PR (which was merged) and some feedback from other reviewers. @jcmgray does the merge of #23588 change anything here?

**Comment by jcmgray** [comment]
Ah yes I hadn't noticed that #23588 had been merged. I think with that the only performance regression I found should be fixed. I'll try and get some time to rebase this and re-test.

For what its worth this implementation has been used by [`cotengra`](https://github.com/jcmgray/cotengra) and [`quimb`](https://github.com/jcmgray/quimb) for the `numpy` backend for a while now without problems.

**Comment by jcmgray** [comment]
As hoped, with https://github.com/numpy/numpy/issues/23588 the following no longer has any slowdown (and any logic about inputs contiguity is deferred to `matmul`):
```python
import numpy as np
d = 1000
x = np.random.randn(d, d)
y = np.random.randn(d, d) + 1j * np.random.randn(d, d)

%%timeit
z = np.einsum('ij,jk->ik', x, y.real, optimize=True)
```
I have also updated all the benchmarks above - the results are pretty much the same. From my perspective this is ready, but still happy to add more test contractions.

**Comment by mattip** [comment]
Is there a larger einsum test suite somewhere we could use to test this out and maybe get some additional benchmarks, or is the existing NumPy testing sufficient?

**Comment by mattip** [comment]
close/reopen to regenerate the merge target with pinning for click

**Comment by jcmgray** [comment]
I couldn't find such a benchmark set that included general contractions (not just matrix multiply equivalent), so made a random pairwise contraction generator and generated a set, the code can be found here: https://github.com/jcmgray/einbench. 

- It allows one to target contraction cost.
- the dataset is half matmul compatible, half with batches indices
- trivially reduceable indices such as sum, trace and diagonal indices are not used, but scalar and outer products are.

Here's the results I get using 4 cores of a AMD EPYC 7742 and openblas. Again, 
- *bmm* is the new method where all calls are dispatched to BLAS. This includes the overhead of the `optimize=True` parsing, which I have improved slightly.
- *dot* is the old method where (most) matmul equivalent calls are dispatched to BLAS. Non matmul-able contractions fall back to `c_einsum`. This includes the overhead of the `optimize=True` parsing.
- *einsum* is the `optimize=False` base `c_einsum` performance (always single threaded).

The data includes each contraction done with: float32, float64, complex64, complex128.

Median GLFOPS binned over target contraction cost:

bench_dataset_gflops

Raw speedup vs `c_einsum`:

bench_dataset_speedup_raw_vs_einsum

Histogram of bmm speedup vs dot:

bench_dataset_speedup_hist_vs_dot

File: numpy/_core/einsumfunc.py [hunk 1]
```diff
@@ -1046,6 +922,312 @@ def einsum_path(*operands, optimize='greedy', einsum_call=False):
...
+    eq : str
+        The einsum equation.
+    a : array_like
+        The first array to contract.
+    b : array_like
+        The second array to contract.
+
+    Returns
+    -------
+    array_like
```
Comment (mattip) [inline numpy/_core/einsumfunc.py:1165]: Maybe add a `Note` section with a reference to https://github.com/jcmgray/einsum_bmm?tab=readme-ov-file#einsum_bmm or to another description of the algorithm?

**Review by mattip — DISMISSED** [review]
This seems like a clear win from the data presented. I think we should put it in.

**Comment by mattip** [comment]
FWIW, I benchmarked this with `spin bench --compare HEAD~1 HEAD -t linalg` and the results are all faster.
<details><summary>Benchmark results></summary>

| Change   | Before [e8687a3b] <einsum_bmm~1>   | After [aef91705] <einsum_bmm>   |   Ratio | Benchmark (Parameter)                                                             |
|----------|------------------------------------|---------------------------------|---------|-----------------------------------------------------------------------------------|
| -        | 28.2±0.2μs                         | 26.7±0.3μs                      |    0.95 | bench_linalg.Einsum.time_einsum_noncon_multiply(<class 'numpy.float32'>)          |
| -        | 33.1±0.05μs                        | 31.4±0.3μs                      |    0.95 | bench_linalg.Einsum.time_einsum_noncon_multiply(<class 'numpy.float64'>)          |
| -        | 485±5ns                            | 460±5ns                         |    0.95 | bench_linalg.LinAlgTransposeVdot.time_transpose((16, 16), 'int64')                |
| -        | 28.3±0.3μs                         | 26.5±0.2μs                      |    0.94 | bench_linalg.Einsum.time_einsum_multiply(<class 'numpy.float32'>)                 |
| -        | 33.1±0.08μs                        | 30.9±0.3μs                      |    0.94 | bench_linalg.Einsum.time_einsum_multiply(<class 'numpy.float64'>)                 |
| -        | 24.0±0.3μs                         | 22.5±0.4μs                      |    0.94 | bench_linalg.Einsum.time_einsum_sum_mul(<class 'numpy.float32'>)                  |
| -        | 7.83±0.4ms                         | 7.40±0.06ms                     |    0.94 | bench_linalg.Linalg.time_pinv('complex64')                                        |
| -        | 65.6±1μs                           | 60.9±0.2μs                      |    0.93 | bench_linalg.Einsum.time_einsum_contig_outstride0(<class 'numpy.float64'>)        |
| -        | 16.0±0.2μs                         | 14.4±0.05μs                     |    0.9  | bench_linalg.Einsum.time_einsum_noncon_mul(<class 'numpy.float64'>)               |
| -        | 16.5±0.1μs                         | 14.7±0.09μs                     |    0.89 | bench_linalg.Einsum.time_einsum_noncon_mul(<class 'numpy.float32'>)               |
| -        | 40.4±0.6μs                         | 34.5±0.6μs                      |    0.85 | bench_linalg.Einsum.time_einsum_contig_outstride0(<class 'numpy.float32'>)        |
| -        | 28.9±0.2μs                         | 24.1±0.09μs                     |    0.84 | bench_linalg.Einsum.time_einsum_noncon_sum_mul2(<class 'numpy.float32'>)          |
| -        | 29.1±0.1μs                         | 23.7±0.1μs                      |    0.82 | bench_linalg.Einsum.time_einsum_noncon_sum_mul(<class 'numpy.float32'>)           |
| -        | 56.3±0.2μs                         | 41.8±0.7μs                      |    0.74 | bench_linalg.Einsum.time_einsum_contig_contig(<class 'numpy.float32'>)            |
| -        | 8.07±0.3ms                         | 5.58±0.4ms                      |    0.69 | bench_linalg.Einsum.time_einsum_noncon_outer(<class 'numpy.float64'>)             |
| -        | 248±9μs                            | 166±8μs                         |    0.67 | bench_linalg.Einsum.time_einsum_mul(<class 'numpy.float32'>)                      |
| -        | 101±1μs                            | 60.1±0.4μs                      |    0.6  | bench_linalg.Einsum.time_einsum_contig_contig(<class 'numpy.float64'>)            |
| -        | 21.6±0.1ms                         | 12.8±0.2ms                      |    0.59 | bench_linalg.Einsum.time_einsum_outer(<class 'numpy.float64'>)                    |
| -        | 12.2±0.2μs                         | 7.06±0.07μs                     |    0.58 | bench_linalg.Einsum.time_einsum_noncon_contig_outstride0(<class 'numpy.float64'>) |
| -        | 12.0±0.1μs                         | 6.80±0.05μs                     |    0.57 | bench_linalg.Einsum.time_einsum_noncon_contig_outstride0(<class 'numpy.float32'>) |
| -        | 182±10μs                           | 101±9μs                         |    0.56 | bench_linalg.Einsum.time_einsum_mul(<class 'numpy.float64'>)                      |
| -        | 11.0±0.4ms                         | 6.04±0.07ms                     |    0.55 | bench_linalg.Einsum.time_einsum_outer(<class 'numpy.float32'>)                    |

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.

</details>

**Comment by mattip** [comment]
@dgasmith would you be available to take a look?

**Comment by jcmgray** [comment]
Great, I'll add a algorithmic note. 

---

I thought I'd also give some extra analysis of the speedups that might be useful for future changes or discussion. The question is basically whether the (rare) cases where the bmm approach is slower that `c_einsum` (other than very small sizes) can be systematically identified. One thing you can look at is arithmetic intensity, here defined by `intensity = prod(dimensions_of_all_involved_indices) / (size_a + size_b + size_output)`. 

- high intensity = matmul like
- low intensity = outer / inner / hadamard product like

If you have a batched matmul with dimensions `(B i j) (B j k) -> (B i k)` then intensity is maximized when `i == j == k`.

If you compare that with speedup with a 'soft cutoff' of 1ms `speedup = (time_einsum + 0.001) / (time_bmm + 0.001)` (since we don't care too much about very small contractions) and bin the same set of contractions above you get a plot like this:

Untitled

In other words:
- intensity is a pretty big predictor for speedup
- all the cases which perform worse have low intensity

It might be tempting to special case low intensity, however, if you plot the mean speedup:

Untitled

Then it is always (just) above 1, so the weight of cases where *low intensity is still sped up by the bmm approach* just about wins out (and includes important cases like `eq = "Bij,Bjk->Bik"` with i, j and k equal but small).

So for the moment, I don't think are super easy ways to customize the dispatch logic. One would probably also have to think about secondary factors like the cost of transposing and reshaping.

**Comment by mattip** [comment]
Something changed in the latest iteration, the `bench_linalg.Einsum.time_einsum_noncon_contig_contig(<class 'numpy.float64'>)` benchmark is now slower than `main`. Is this expected?
```
| Change   | Before [3abf5795] <einsum_bmm~2>   | After [d2c1ccfc] <einsum_bmm>   |   Ratio | Benchmark (Parameter)                                                             |
|----------|------------------------------------|---------------------------------|---------|-----------------------------------------------------------------------------------|
| +        | 16.9±0.2μs                         | 17.8±0.2μs                      |    1.06 | bench_linalg.Einsum.time_einsum_noncon_contig_contig(<class 'numpy.float64'>)     |
| -        | 11.0±0.2ms                         | 10.4±0.2ms                      |    0.95 | bench_linalg.MatmulStrided.time_matmul('matmul_m400_p400_n100_bs01')              |
| -        | 3.12±0.04μs                        | 2.94±0.06μs                     |    0.94 | bench_linalg.LinalgSmallArrays.time_det_small_array                               |
| -        | 28.7±0.1μs                         | 26.5±0.3μs                      |    0.92 | bench_linalg.Einsum.time_einsum_multiply(<class 'numpy.float32'>)                 |
| -        | 25.5±0.2μs                         | 23.4±0.1μs                      |    0.92 | bench_linalg.Einsum.time_einsum_sum_mul2(<class 'numpy.float32'>)                 |
| -        | 28.8±0.07μs                        | 26.2±0.1μs                      |    0.91 | bench_linalg.Einsum.time_einsum_noncon_multiply(<class 'numpy.float32'>)          |
| -        | 33.5±0.2μs                         | 30.4±0.5μs                      |    0.91 | bench_linalg.Einsum.time_einsum_noncon_multiply(<class 'numpy.float64'>)          |
| -        | 25.3±0.3μs                         | 23.0±0.6μs                      |    0.91 | bench_linalg.Einsum.time_einsum_sum_mul(<class 'numpy.float32'>)                  |
| -        | 33.6±0.2μs                         | 30.2±0.8μs                      |    0.9  | bench_linalg.Einsum.time_einsum_multiply(<class 'numpy.float64'>)                 |
| -        | 66.9±0.2μs                         | 59.5±1μs                        |    0.89 | bench_linalg.Einsum.time_einsum_contig_outstride0(<class 'numpy.float64'>)        |
| -        | 7.25±0.5ms                         | 6.49±0.05ms                     |    0.89 | bench_linalg.Linalg.time_pinv('complex128')                                       |
| -        | 1.51±0.05ms                        | 1.33±0.01ms                     |    0.88 | bench_linalg.Einsum.time_einsum_noncon_outer(<class 'numpy.float32'>)             |
| -        | 17.7±0.4μs                         | 15.0±0.1μs                      |    0.85 | bench_linalg.Einsum.time_einsum_noncon_mul(<class 'numpy.float32'>)               |
| -        | 16.5±0.03μs                        | 14.0±0.1μs                      |    0.85 | bench_linalg.Einsum.time_einsum_noncon_mul(<class 'numpy.float64'>)               |
| -        | 40.9±0.1μs                         | 34.2±0.5μs                      |    0.84 | bench_linalg.Einsum.time_einsum_contig_outstride0(<class 'numpy.float32'>)        |
| -        | 28.6±0.1μs                         | 23.8±0.7μs                      |    0.83 | bench_linalg.Einsum.time_einsum_noncon_sum_mul(<class 'numpy.float32'>)           |
| -        | 29.1±0.05μs                        | 24.1±0.2μs                      |    0.83 | bench_linalg.Einsum.time_einsum_noncon_sum_mul2(<class 'numpy.float32'>)          |
| -        | 56.2±0.2μs                         | 43.5±0.7μs                      |    0.77 | bench_linalg.Einsum.time_einsum_contig_contig(<class 'numpy.float32'>)            |
| -        | 264±10μs                           | 173±8μs                         |    0.65 | bench_linalg.Einsum.time_einsum_mul(<class 'numpy.float32'>)                      |
| -        | 8.11±0.3ms                         | 5.10±0.2ms                      |    0.63 | bench_linalg.Einsum.time_einsum_noncon_outer(<class 'numpy.float64'>)             |
| -        | 97.7±1μs                           | 58.6±1μs                        |    0.6  | bench_linalg.Einsum.time_einsum_contig_contig(<class 'numpy.float64'>)            |
| -        | 12.2±0.1μs                         | 7.34±0.2μs                      |    0.6  | bench_linalg.Einsum.time_einsum_noncon_contig_outstride0(<class 'numpy.float32'>) |
| -        | 12.2±0.1μs                         | 7.13±0.2μs                      |    0.58 | bench_linalg.Einsum.time_einsum_noncon_contig_outstride0(<class 'numpy.float64'>) |
| -        | 10.6±0.3ms                         | 5.95±0.05ms                     |    0.56 | bench_linalg.Einsum.time_einsum_outer(<class 'numpy.float32'>)                    |
| -        | 22.5±0.4ms                         | 12.7±0.09ms                     |    0.56 | bench_linalg.Einsum.time_einsum_outer(<class 'numpy.float64'>)                    |
| -        | 184±20μs                           | 102±5μs                         |    0.55 | bench_linalg.Einsum.time_einsum_mul(<class 'numpy.float64'>)                      |

```

**Comment by jcmgray** [comment]
I haven't changed any code since the previous benchmarks (apologies, I didn't really think about the rebase and force push). 

For `time_einsum_noncon_contig_contig` the equation is "ji,i->" with shapes (30, 40) and (40,) and non-contiguous arrays. The prior approach will parse the equation, and then dispatch the pair to `c_einsum` anyway. The new approach will simplify the left array `"ji->i"` then perform `"i,i->"` via matmul. So they are doing different things, but happen to benchmark very similarly at these sizes. It's not unexpected that the new approach might genuinely be slightly slower at small sizes. But notably, if the sizes increase the new approach starts to have a speedup.

Locally I get: 
shapes: (30, 40), (40,)
OLD: **14.5** μs ± 157 ns | NEW: **15.8** μs ± 82.9 ns
shapes: (300, 400), (400,) 
OLD: **118** μs ± 1.34 μs | NEW: **74.4** μs ± 454 ns

Generally at the 10-20us level a significant part of the time is just the equation being parsing, which should also on average be slightly faster now, but was not the main thrust of this PR.

**Comment by mattip** [comment]
Thanks @jcmgray. Let's get this in. There is some cleanup (typos, tweak cache sizes) but I don't want to bikeshed on this too hard. We can clean that up in subsequent PRs. The code itself seems solid, and the improvements for large arrays are impressive.

**Comment by jcmgray** [comment]
Great, thanks for the review @mattip. Happy to help on further cleanup.

**Comment by lucascolley** [comment]
I am suspicious that this PR has caused fallout for the SciPy test suite — see for example the failures in the dev deps job of [this log](https://github.com/scipy/scipy/actions/runs/19393070829/job/55489304653), where the traceback is showing

```
/opt/hostedtoolcache/Python/3.14.0/x64/lib/python3.14/site-packages/numpy/_core/einsumfunc.py:1639: in einsum
    new_view = bmm_einsum(einsum_str, *tmp_operands, **kwargs)
```

The failures are all restricted to `scipy/stats/tests/test_multivariate.py::TestMatrixT`, so it may be that the changes here have just exposed a user error. Perhaps you could take a look @jcmgray ?

**Comment by mattip** [comment]
It seems there is a problem in the parsing code. Here is a stand-alone reproducer taken from the test failure. Note `optmize=False` works, but `optimize=True` prints an error
<details></summarize>test.py</summarize>

```
import numpy as np

rowchol = np.array([[[0.40373579]], [[0.59136425]], [[0.68318971]]])
std_norm = np.array([[[ 0.49671415,  1.52302986,  1.57921282]],
       [[-0.1382643 , -0.23415337,  0.76743473]],
       [[ 0.64768854, -0.23413696, -0.46947439]]])
colchol    = np.array([[[1.        , 0.        , 0.        ],
        [0.3       , 0.9539392 , 0.        ],
        [0.2       , 0.35641684, 0.91267028]]])

t_raw = np.einsum("ijp,ipq,ikq->ijk", rowchol, std_norm, colchol, optimize=False)
t_raw = np.einsum("ijp,ipq,ikq->ijk", rowchol, std_norm, colchol, optimize=True)
```

</details>

Running the script prints this error
```
/tmp/venv313/bin/python3.13 -P /tmp/issue23513.py
Traceback (most recent call last):
  File "/tmp/issue23513.py", line 12, in <module>
    t_raw = np.einsum("ijp,ipq,ikq->ijk", rowchol, std_norm, colchol, optimize=True)
  File "/home/matti/oss/numpy/build-install/usr/lib/python3/dist-packages/numpy/_core/einsumfunc.py", line 1639, in einsum
    new_view = bmm_einsum(einsum_str, *tmp_operands, **kwargs)
  File "/home/matti/oss/numpy/build-install/usr/lib/python3/dist-packages/numpy/_core/einsumfunc.py", line 1180, in bmm_einsum
    ) = _parse_eq_to_batch_matmul(eq, a.shape, b.shape)
        ~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "/home/matti/oss/numpy/build-install/usr/lib/python3/dist-packages/numpy/_core/einsumfunc.py", line 1113, in _parse_eq_to_batch_matmul
    perm_ab = tuple(out_produced.index(ix) for ix in out)
  File "/home/matti/oss/numpy/build-install/usr/lib/python3/dist-packages/numpy/_core/einsumfunc.py", line 1113, in <genexpr>
    perm_ab = tuple(out_produced.index(ix) for ix in out)
                    ~~~~~~~~~~~~~~~~~~^^^^
ValueError: substring not found
```

**Comment by jcmgray** [comment]
Thanks for the reproducer, I will take a look shortly.

**Comment by jcmgray** [comment]
This should be fixed in #30242. It was a genuine parsing error encountered when broadcasting certain "singleton" indices -> size 1 on one array, != 1 on the other. Apologies, there are so many edge cases in einsum!

I've now added tests to catch this case (and also locally tested on a further 1,000,000 randomly generated equations).

**Comment by hawkinsp** [comment]
I realize that `einsum_call` is a hidden argument to `einsum_path`, but while looking at breakages from NumPy 2.4.0rc1 in our code base I found that there is code in the wild that relies on it and will be broken by NumPy 2.4: https://github.com/pyscf/pyscf/blob/8207b63d3d48acaebebd3487f07b91e4eef1fcab/pyscf/lib/numpy_helper.py#L243

Is it worth adding this to the NumPy release notes, at least?

**Comment by jcmgray** [comment]
I don’t think this PR changed anything to do with ‘einsum_call’. Do you maybe have a link to the failing tests to see what the error is?

**Comment by jcmgray** [comment]
Oh I see, the number of items in the `contraction_list` returned has changed.

**Comment by jcmgray** [comment]
So the issue is this PR changed the elements of the `contraction_list` returned when the hidden kwarg `einsum_call=True` from:
```python
inds, idx_rm, einsum_str, remaining, blas = contraction
```
to
```python
inds, einsum_str, remaining = contraction
```
since `idx_rm` and `blas` aren't needed anymore (and aren't computed). But various libraries do seem to rely on using this internal api, generally for accessing the sequence of pairwise `einsum_str` only. I guess the options are:

1. leave as is since it is a hidden option, try and inform user libraries
2. return e.g. `inds, None, einsum_str, remaining, None`, so that libraries that just access `einsum_str` from the tuple don't need to change anything
3. try and maintain exact returned form of `einsum_call=True` and use e.g. `einsum_call="bmm"` for the actual internal numpy call.

**Comment by mattip** [comment]
> Is it worth adding this to the NumPy release notes, at least?

I think we can add a release note, and the downstream libraries can adjust, using `if numpy.__version__.split('.') >= ['2', '4']`. I am curious if the technique used in pyscf actually speeds up special cases of `einsum` and if so should we should adopt it?

**Comment by hawkinsp** [comment]
I have no strong opinion, personally, it just surprised me that it was an apparently breaking change that wasn't in the release notes.

**Comment by jcmgray** [comment]
> I am curious if the technique used in pyscf actually speeds up special cases of einsum and if so should we should adopt it?

Pyscf has a few special cases for pairwise contraction:

1. if `tblis` is installed (https://github.com/pyscf/pyscf-tblis) use it for double dtype contractions. This can often be a bit faster than transpose+matmul.
2. for *non* batch matmul compatible, explicitly do the transposes, make sure contiguous, and call the blas wrappers directly, this should be similar to `bmm_einsum` or the previous optimize=True routine.
3. if there are mixed dtypes (i.e. complex and real) compute the real and imag parts separately. I guess there is a slight advantage to doing this.

If no `tblis`, and the contraction isn't matmul like, or is small, it falls back to numpy.einsum.
